ls1intum/ArTEMiS
# Feature/stager approach

### Checklist

### Motivation and Context
To verify code quality or evaluate creative programming tasks, the manual review of code submissions is necessary. The process identifying of the students’ code, their contributions, and assessing their solution can require many repetitive manual steps.

### Description

This implementation automatically:

1. Downloads repositories (already implemented)
2. Filter late submissions (commits after programming exercise due date)
3. Combines student commits (after last instructor commit)

The downloaded repositories will have an additional **local only** `stager` branch (s. image attached).
The origin repo is *left untouched at all times* so tutors can always see the full commit history.

### Steps for Testing


1. Log in to ArTEMiS
2. Navigate to Course Administration
3. Create a new exercise with template, solution, build plan
4. Assign students to exercise and randomly add some commits to their repositories
5. Download repos as zip file from: `/#/course/{id}/exercise/{id}/dashboard`
6. Verify that the downloaded repositories have a correct `stager` branch
7. No exceptions are thrown during the process

### Screenshots

![download](https://user-images.githubusercontent.com/7462542/56820619-b54c7200-684c-11e9-8605-b0117131b640.jpg)
![sourcetree](https://user-images.githubusercontent.com/7462542/56820620-b54c7200-684c-11e9-9297-d1ff4ac9cf24.jpg)

### Next steps (new PR)

1. Normalize code style / ignore whitespace --> [JGIT as blocker](https://bugs.eclipse.org/bugs/show_bug.cgi?id=471845)
2. Rename projects --> maybe in BambooService after `BAMBOO_EMPTY_COMMIT_WORKAROUND_NECESSARY`. Ask others first about this step.
3. Evaluate if `GitService` should support multiple folders instead of one `REPO_CLONE_PATH`
